Blog: Using MQTT Sparkplug B with Node-RED (#2478)
FlowFuse/website: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md

---
title: Using MQTT Sparkplug B with Node-RED
subtitle: A Practical Guide to integrating MQTT Sparkplug B with Node-RED
description: Explore how MQTT Sparkplug B enhances MQTT with standardized data formats, making integration easier. Learn how to set up Sparkplug B in Node-RED for efficient data management in industrial environments.
date: 2024-08-21
authors: ["sumit-shinde"]

> image: 

Reviewer: @sumitshinde-84 Please make a tile

Author: Done !

Author: we need to replace the diagrams !

Reviewer 2: > we need to replace the diagrams !

Are you saying the article wasnt 100% ready @sumitshinde-84?

FWIW, i typically switch a PR to draft if not ready to go.

Author: The article is ready; it just needs the diagram updated with the diagrams Yndira will recreate.

Author: I created those rough diagrams so that Yndira can understand the type of diagram we need.

tags:
   - post
   - nodered
   - mqtt sparkplug b with node-red
---

> Connected devices are creating a lot of data, but without a standard format, managing this data can be tricky. MQTT helps with messaging, but it can be inconsistent. MQTT Sparkplug B solves this problem by providing a clear, standardized format for data. In this guide, we’ll show you how to use MQTT Sparkplug B with Node-RED to make managing your device data easier and more organized.

Reviewer: ```suggestion
Connected devices are creating a lot of data but without a standard format, managing this data can be tricky. MQTT helps with messaging, though MQTT is flexible and doesn't standardize message formats. MQTT Sparkplug B solves this problem by providing a clear, standardized format for data. In this guide, we’ll show you how to use MQTT Sparkplug B with Node-RED to make managing your device data easier and more organized.
```

Reviewer 2: NB: A counter suggestion was offered in my review here: https://github.com/FlowFuse/website/pull/2478#discussion_r1724909709

Reviewer 2: outdated

<!--more-->

## What is MQTT Sparkplug?

> *Sparkplug B* is an open-source specification hosted at the Eclipse Foundation that extends MQTT by providing a standardized way to structure data. It ensures that all devices use a consistent message format, that simplifies integration and improves data consistency.

Reviewer: I think a key point is that it's optimized in terms of structure for sensor readings and device data being published. It's not a standard for pushing changes to a device or sensor.

Reviewer 2: NB: A counter suggestion was offered in my review here: https://github.com/FlowFuse/website/pull/2478#discussion_r1724915459

Reviewer 2: outdated

> *Sparkplug B* is an open-source specification hosted at the Eclipse Foundation that extends MQTT by providing a standardized way to structure data. It ensures that all devices use a consistent message format, that simplifies integration and improves data consistency.

Reviewer 2: "improves data consistency" didnt "feel right"

In terms of data, consistency is often related to timing and factor not necessarily related to "structure".

> *Sparkplug B* is an open-source specification hosted at the Eclipse Foundation that extends MQTT by providing a standardized way to structure data. It ensures that all devices use a consistent message format, that simplifies integration and improves data consistency.

Reviewer 2: This re-write removes the "simplifies things" and sticks to the official line from ecplise.  the "simplifies things" is touched on in the next sentence so this was somewhat superfluous IMHO.

- **Sequence Number**: Each message includes a sequence number that increments with each new message. This helps detect any missing messages and ensures data consistency.

### Topic Naming Conventions

To effectively manage and route messages in Sparkplug B, it uses a well-structured topic namespace. This structured approach ensures that messages are organized logically and can be easily identified and processed by different systems

The topic namespace in Sparkplug B follows a specific format:

```
spBv1.0/[Group ID]/[Message Type]/[EON Node ID]/[Device ID]
```

- **[Group ID]**: Represents the group or application context.
- **[Message Type]**: Specifies the type of message (e.g., NBIRTH, NDATA).
- **[EON Node ID]**: Identifies the edge node or system.
- **[Device ID]**: Identifies the specific device (if applicable).

## Using MQTT Sparkplug with Node-RED

In this section, we'll explore an example where Machine1 on the factory floor sends temperature and humidity data to an MQTT broker. We’ll cover how to integrate MQTT Sparkplug B into Node-RED and demonstrate basic operations such as sending and receiving data using MQTT Sparkplug B and more.

> ### Prequsite

Reviewer 2: ```suggestion
### Prerequisite
```

Before you begin, ensure you have the following:

> - [node-red-contrib-mqtt-sparkplug-plus](https://flows.nodered.org/node/node-red-contrib-mqtt-sparkplug-plus): Install this Node-RED package for Sparkplug B support via pallete manager.

Reviewer 2: ```suggestion
- [node-red-contrib-mqtt-sparkplug-plus](https://flows.nodered.org/node/node-red-contrib-mqtt-sparkplug-plus): Install this Node-RED package for Sparkplug B support via palette manager.
```

    ```json
    {
        "metrics": [
            {
                "name": "sensor/temperature",
                "value": $random() * 100
            },
            {
                "name": "sensor/humidity",
                "value": $random() * 100
            }
        ]
    }
    ```

2. Drag the mqtt sparkplug device node onto the canvas.
3. Double-click the mqtt sparkplug device node to open the configuration panel. Add the metric names that you will be sending by clicking the bottom-left "Add" button. Ensure that the names match the metric names in the payload you are sending and specify the data types for each metric.

![Screenshot showing the Sparkplug Device node configuration and the "Add" button for defining metrics](./images/mqtt-sparkplug-device-node.png "Screenshot showing the Sparkplug Device node configuration and the 'Add' button for defining metrics")  
_Screenshot showing the Sparkplug Device node configuration and the "Add" button for defining metrics_

4. Switch to the Advanced tab by clicking the "Advanced" option at the top-right.

> 5. Enable the "Send Birth Immediately" option. This ensures that a Birth message (DBIRTH) is sent immediately upon deployment and connection to the MQTT broker. Note that enabling this option will send the `DBIRTH` message when the device node connects, but an `NBIRTH` message will be sent successfull connection of mqtt sparkplug out node if you are using.

Reviewer 2: ```suggestion
5. Enable the "Send Birth Immediately" option. This ensures that a Birth message (DBIRTH) is sent immediately upon deployment and connection to the MQTT broker. Note that enabling this option will send the `DBIRTH` message when the device node connects, but an `NBIRTH` message will be sent successful connection of mqtt sparkplug out node if you are using.
```

![Screenshot showing the Sparkplug Device node configuration and the "Add" button for defining metrics](./images/mqtt-spark-device-advance.png "Screenshot showing the Sparkplug Device node configuration and the 'Add' button for defining metrics")  
_Screenshot showing the Sparkplug Device node configuration and the "Add" button for defining metrics_

6. Optionally, enable Store and Forward when not connected to ensure that messages are stored and sent once the connection is re-established. To use this option, make sure you have enabled it in the mqtt sparkplug broker config node and specified the destination.
7. Connect the inject node's output to the mqtt sparkplug device node's input.
8. Deploy the flow by clicking the top-right "Deploy" button.

> Once you deploy the flow and all devices connect to the MQTT broker, the system automatically send a `DBIRTH` message as soon as each device within the node connects, signaling that the device is ready for data transmission.

Reviewer 2: ```suggestion
Once you deploy the flow and all devices connect to the MQTT broker, the system automatically send a `DBIRTH` message as soon as each device within the node connects, signalling that the device is ready for data transmission.
```
